bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/template/RuleTemplate.java

```diff
@@ -117,13 +116,13 @@ public RuleTemplate(@Nullable String UID, @Nullable String label, @Nullable Stri
         this.uid = UID == null ? UUID.randomUUID().toString() : UID;
         this.label = label;
         this.description = description;
-        this.triggers = triggers == null ? Collections.emptyList() : Collections.unmodifiableList(triggers);
-        this.conditions = conditions == null ? Collections.emptyList() : Collections.unmodifiableList(conditions);
-        this.actions = actions == null ? Collections.emptyList() : Collections.unmodifiableList(actions);
-        this.configDescriptions = configDescriptions == null ? Collections.emptyList()
+        this.triggers = triggers == null ? List.of() : Collections.unmodifiableList(triggers);
+        this.conditions = conditions == null ? List.of() : Collections.unmodifiableList(conditions);
+        this.actions = actions == null ? List.of() : Collections.unmodifiableList(actions);
+        this.configDescriptions = configDescriptions == null ? List.of()
                 : Collections.unmodifiableList(configDescriptions);
         this.visibility = visibility == null ? Visibility.VISIBLE : visibility;
-        this.tags = tags == null ? new HashSet<>() : new HashSet<>(tags);
+        this.tags = tags == null ? Set.of() : Collections.unmodifiableSet(tags);
```

Comment:
Why do you use `Collections.unmodifiableSet` instead of `Set.copyOf`?

Reply:
I made it similar to the other code that uses `unmodifiableList`. Using `unmodifiableList` has some benefits like better performance because it doesn't allocate a new list and it is "read through". But that can also give issues when the code providing the input parameters updates the collections.